File: General Testing/Controlsystem/Controller.py
import time
import VelocityProfile
# from cameraComputeDC import camera_get_dc


class ThrottleController:
    def __init__(self, kp, ki, kd):
        self.Kp = kp
        self.Ki = ki
        self.Kd = kd
        self.pwm = 15.0
        self.xd = 0
        self.vd = 0
        self.x = 0
        self.x_previous = 0
        self.v = 0
        self.last_update = 0  # Initialize with 0
        self.delta_t = 0  # Initialize delta_t
        self.position_error = 0
        self.integral_error = 0
        self.derivative_error = 0
        self.velocity_profile = VelocityProfile.VelocityProfile



    def update_pwm(self, start_time, velocity_profile):
        current_time = time.time()  # Convert to milliseconds
        # Get current position
        # self.x = self.readEncoders        # RESTORE THIS
        self.x = 2
        #Calculate current position

        # Get desired position
        self.xd = velocity_profile.get_desired_position(start_time)
        # Calculate delta t
        self.delta_t = (time.time() - self.last_update) 
        # Calculate velocity
        if self.delta_t != 0:
            print(self.x)
            print(self.x_previous)
            self.v = (self.x - self.x_previous) / self.delta_t
        #Get desired velocity
        self.vd = velocity_profile.get_desired_velocity(start_time)
        # Calculate Errors
        self.position_error = self.xd - self.x
        self.integral_error += self.position_error
        self.integral_error = max(-1000, min(1000, self.integral_error)) # Prevent integral windup #Choose different values maybe
        self.derivative_error = self.vd - self.v
        # Calculate next PWM value
        self.pwm = self.Kp * self.position_error + self.Ki * self.integral_error + self.Kd * self.derivative_error
        # Constrain between 10 and 20% DC
        self.pwm = max(10, min(20, self.pwm))  
        
        # Update previous variables
        self.last_update = time.time()
        self.x_previous = self.x
        # Add delay
        time.sleep(0.01)

    def readEncoders(self): # Need to write that
        self.x
        return int(self.x)


    def reset(self):
        self.x = 0
        self.xd = 0
        self.x_previous = 0
        self.v = 0
        self.vd = 0
        self.pwm = 0
        self.position_error = 0
        self.integral_error = 0
        self.derivative_error = 0
        self.last_update = 0
        self.delta_t = 0 



class SteeringController:
    def __init__(self, kp, ki, kd):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.pwm = 15.0
        self.delta_t = 0
        self.last_updated = 0
        self.desired_dist_to_centerline = 0
        self.current_dist_to_centerline = 0
        self.previous_dist_to_centerline = 0
        self.error_dist_to_centerline = 0
        self.derivative_error_dist_to_centerline = 0
        self.integral_error_dist_to_centerline = 0
        
    def update_pwm(self, desired_dist_to_centerline, actual_dist_to_center):
        # Get current time 
        current_time = time.time()
        # Get desired steering angle
        self.desired_dist_to_centerline = desired_dist_to_centerline
        # Get current steering angle (m and b from MATLAB file "CarSteeringModel" relating PWM signal and steering angle centerline)   
        #    
        self.current_dist_to_centerline = -1 * actual_dist_to_center     #camera value


        # Update delta T
        self.delta_t = time.time() - self.last_updated
        # Calculate errors
        self.error_dist_to_centerline = self.desired_dist_to_centerline - self.current_dist_to_centerline
        if self.delta_t != 0:
            self.derivative_error_dist_to_centerline = (self.current_dist_to_centerline - self.previous_dist_to_centerline) / self.delta_t
        self.integral_error_dist_to_centerline += self.error_dist_to_centerline
        # Update PWM
        self.pwm = 15 + self.kp * self.error_dist_to_centerline + self.ki * self.integral_error_dist_to_centerline + self.kd * self.derivative_error_dist_to_centerline

        print(f"Integral: {self.ki * self.integral_error_dist_to_centerline}")
        print(f"Proportional: {self.kp * self.error_dist_to_centerline}")
        # Contrain PWM
        self.pwm = max(10,min(self.pwm, 20))
        
        # Update previous variables 
        self.previous_dist_to_centerline = self.current_dist_to_centerline
        self.last_updated = time.time()
        # Add delay
        time.sleep(0.01)

    def reset(self):

        self.pwm = 15
        self.delta_t = 0
        self.last_updated = 0
        self.desired_dist_to_centerline = 0
        self.current_dist_to_centerline = 0
        self.previous_dist_to_centerline = 0
        self.error_dist_to_centerline = 0
        self.derivative_error_dist_to_centerline = 0
        self.integral_error_dist_to_centerline = 0

File: Car Testing/Controlsystem/Controller.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h> // For usleep function

typedef struct {
    double Kp, Ki, Kd;
    double pwm;
    double xd, vd;
    double x, x_previous;
    double v;
    double last_update;
    double deltaT;
    double position_error, integral_error, derivative_error;
} ThrottleController;

typedef struct {
    double kp, ki, kd;
    double pwm;
    double deltaT, last_updated;
    double desired_steering_angle;
    double current_steering_angle;
    double previous_steering_angle;
    double steering_angle_error;
    double steering_angle_derivative_error;
    double steering_angle_integral_error;
} SteeringController;

// Function prototypes
double get_desired_position(double start_time);
double get_desired_velocity(double start_time);
double readEncoders();
double get_current_time();

void throttle_update_pwm(ThrottleController *tc, double start_time) {
    double current_time = get_current_time();
    
    tc->x = readEncoders();
    tc->xd = get_desired_position(start_time);
    tc->deltaT = current_time - tc->last_update;

    if (tc->deltaT != 0)
        tc->v = (tc->x - tc->x_previous) / tc->deltaT;

    tc->vd = get_desired_velocity(start_time);
    tc->position_error = tc->xd - tc->x;
    tc->integral_error += tc->position_error;

    // Prevent integral windup
    if (tc->integral_error > 1000) tc->integral_error = 1000;
    if (tc->integral_error < -1000) tc->integral_error = -1000;

    tc->derivative_error = tc->vd - tc->v;
    tc->pwm = tc->Kp * tc->position_error + tc->Ki * tc->integral_error + tc->Kd * tc->derivative_error;

    // Constrain between 10 and 20% DC
    if (tc->pwm > 20) tc->pwm = 20;
    if (tc->pwm < 10) tc->pwm = 10;

    tc->last_update = get_current_time();
    tc->x_previous = tc->x;
    
    usleep(10000); // 10 ms delay
}

void throttle_reset(ThrottleController *tc) {
    tc->x = tc->xd = tc->x_previous = 0;
    tc->v = tc->vd = 0;
    tc->pwm = 0;
    tc->position_error = tc->integral_error = tc->derivative_error = 0;
    tc->last_update = 0;
    tc->deltaT = 0;
}

void steering_update_pwm(SteeringController *sc, double desired_steeringAngle, double current_steering_angle) {
    double current_time = get_current_time();
    
    sc->desired_steering_angle = desired_steeringAngle;
    sc->current_steering_angle = current_steering_angle;
    sc->deltaT = current_time - sc->last_updated;
    sc->steering_angle_error = sc->desired_steering_angle - sc->current_steering_angle;

    if (sc->deltaT != 0)
        sc->steering_angle_derivative_error = (sc->current_steering_angle - sc->previous_steering_angle) / sc->deltaT;

    sc->steering_angle_integral_error += sc->steering_angle_error;
    sc->pwm = sc->kp * sc->steering_angle_error + sc->ki * sc->steering_angle_integral_error + sc->kd * sc->steering_angle_derivative_error;

    // Constrain PWM
    if (sc->pwm > 17) sc->pwm = 17;
    if (sc->pwm < 13) sc->pwm = 13;

    sc->previous_steering_angle = sc->current_steering_angle;
    sc->last_updated = get_current_time();
    
    usleep(10000); // 10 ms delay
}

void steering_reset(SteeringController *sc) {
    sc->pwm = 15;
    sc->deltaT = sc->last_updated = 0;
    sc->desired_steering_angle = sc->current_steering_angle = sc->previous_steering_angle = 0;
    sc->steering_angle_error = sc->steering_angle_derivative_error = sc->steering_angle_integral_error = 0;
}

// Dummy implementations for external functions
double get_desired_position(double start_time) {
    return 10.0; // Example value
}

double get_desired_velocity(double start_time) {
    return 2.0; // Example value
}

double readEncoders() {
    return rand() % 100; // Example random encoder value
}

double get_current_time() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

int main() {
    ThrottleController tc = {1.0, 0.1, 0.05, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    SteeringController sc = {1.0, 0.1, 0.05, 15, 0, 0, 0, 0, 0, 0, 0, 0};

    double start_time = get_current_time();

    for (int i = 0; i < 100; i++) {
        throttle_update_pwm(&tc, start_time);
        steering_update_pwm(&sc, 10.0, 5.0);
        printf("Throttle PWM: %.2f, Steering PWM: %.2f\n", tc.pwm, sc.pwm);
        usleep(50000); // 50 ms delay
    }

    return 0;
}
